source/simulator/core.cc

```diff
@@ -2241,6 +2256,297 @@ namespace aspect
+                                      &&
+                                      (i >= parameters.max_nonlinear_iterations_in_prerefinement)))); ++i)
+            {
+              assemble_newton_stokes_system = true;
+
+
+              double stokes_residual = 0;
+
+              solution.block(introspection.block_indices.pressure) = 0;
+              solution.block(introspection.block_indices.velocities) = 0;
+
+              pcout << "Newton iteration number: " << nonlinear_iteration << std::endl;
+
+              assemble_advection_system(AdvectionField::temperature());
+              build_advection_preconditioner(AdvectionField::temperature(),
+                                             T_preconditioner);
+
+              if (nonlinear_iteration == 0)
+                initial_temperature_residual = system_rhs.block(introspection.block_indices.temperature).l2_norm();
+
+              const double temperature_residual = solve_advection(AdvectionField::temperature());
+
+              current_linearization_point.block(introspection.block_indices.temperature)
+                = solution.block(introspection.block_indices.temperature);
+              rebuild_stokes_matrix = true;
+              std::vector<double> composition_residual (parameters.n_compositional_fields,0);
+
+              for (unsigned int c=0; c<parameters.n_compositional_fields; ++c)
+                {
+                  assemble_advection_system (AdvectionField::composition(c));
+                  build_advection_preconditioner(AdvectionField::composition(c),
+                                                 C_preconditioner);
+
+                  if (nonlinear_iteration == 0)
+                    initial_composition_residual[c] = system_rhs.block(introspection.block_indices.compositional_fields[c]).l2_norm();
+
+                  composition_residual[c]
+                    = solve_advection(AdvectionField::composition(c));
+                }
+
+              // for consistency we update the current linearization point only after we have solved
+              // all fields, so that we use the same point in time for every field when solving
+              for (unsigned int c=0; c<parameters.n_compositional_fields; ++c)
+                current_linearization_point.block(introspection.block_indices.compositional_fields[c])
+                  = solution.block(introspection.block_indices.compositional_fields[c]);
+              LinearAlgebra::BlockVector global_backup_linearization_point = current_linearization_point;
+
+              if (use_picard == true && (max <= parameters.nonlinear_switch_tolerance ||
+                                         nonlinear_iteration >= parameters.max_pre_newton_nonlinear_iterations))
+                {
+                  use_picard = false;
+                  pcout << "Switching Newton derivatives on." << std::endl;
+
+                  bool use_said_method = false;
+                  if (!use_said_method)
```

Comment:
I should make this into a global variable, or remove it. My current preference it making it in a global variable, because I suspect that is could stabilize hard computations, and I would like to be able to quickly test that when other methods fail.